test/refract/data/test-Array.cc

```diff
@@ -0,0 +1,512 @@
+                {
+                    REQUIRE(array.begin()[0].get() == mock1ptr);
+                    REQUIRE(array.begin()[1].get() == mock2ptr);
+                    REQUIRE(array.begin()[2].get() == mock3ptr);
+                }
+                THEN("it is iterable")
+                {
+                    auto mocks = std::array<IElement*, 3>{ mock1ptr, mock2ptr, mock3ptr };
+                    auto mocks_it = mocks.begin();
+                    int ctx = 0;
+                    for (const auto& el : array) {
+                        REQUIRE(mocks_it < mocks.end()); // memory overflow
+                        REQUIRE(*mocks_it == el.get());
+                        ++mocks_it;
+                        ++ctx;
+                        REQUIRE(ctx != 0); // integer overflow
+                    }
+                    REQUIRE(ctx == 3);
+                }
+                THEN("non-temporary ElementMocks were not constructed")
+                {
+                    REQUIRE(test::ElementMock::instances().size() == 3);
+                }
+                THEN("nothing was called on the mocks")
+                {
+                    REQUIRE(mock1ptr->_total_ctx == 0);
+                    REQUIRE(mock2ptr->_total_ctx == 0);
+                    REQUIRE(mock3ptr->_total_ctx == 0);
+                }
+
+                WHEN("another array_t is copy constructed from it")
+                {
+                    auto mock1clone = new test::ElementMock{};
+                    auto mock2clone = new test::ElementMock{};
+                    auto mock3clone = new test::ElementMock{};
+                    REQUIRE(test::ElementMock::instances().size() == 6);
+
+                    mock1ptr->clone_out.reset(mock1clone);
+                    mock2ptr->clone_out.reset(mock2clone);
+                    mock3ptr->clone_out.reset(mock3clone);
+                    REQUIRE(test::ElementMock::instances().size() == 6);
+
+                    array_t array2(array);
+
+                    THEN("its size is three")
+                    {
+                        REQUIRE(array2.size() == 3);
+                    }
+                    THEN("its begin is 3 away from its end")
+                    {
+                        REQUIRE(array2.end() - array2.begin() == 3);
+                    }
+                    THEN("the original's size is three")
+                    {
+                        REQUIRE(array.size() == 3);
+                    }
+                    THEN("it is not empty")
+                    {
+                        REQUIRE(!array2.empty());
+                    }
+                    THEN("the original is not empty")
+                    {
+                        REQUIRE(!array.empty());
+                    }
+                    THEN("its size equals the original's size")
+                    {
+                        REQUIRE(array.size() == array2.size());
+                    }
+                    THEN("it is iterable")
```

Comment:
DRY
It is third repetition of  **is iterable** test case with simillar implementation, what about come macro/function?

Reply:
I don't like to generalize in tests. If an assertion fails, I want it to be stated where is belongs. But I will think of a way.